Support unreported capital transactions (Form 8949)

zakpatterson opened the pull request:
Adds a new page to the menu for adding sales of assets not
reported on schedule 1099-B. This can include cryptocurrency,
securities owned since before changes in reporting requirements,
and real estate.

Updates Schedule D to incorporate 8949 data, set up for 2020 and 2021.

Assets live outside of the tax years structure at the root of our state datatype. Assets have optional `closeDate`s, so those assets are selected based on `closeDate` in the current tax year for inclusion on the return.

**What kind of change does this PR introduce?**
- Feature

noahm commented on src/components/income/OtherInvestments.tsx:
```diff
@@ -0,0 +1,198 @@
+        />
+        <DatePicker
+          maxDate={new Date(2021, 12, 31)}
+          label={positionType === 'Real Estate' ? 'Sale date' : 'Close Date'}
```
Just as an average person I think "purchase date" (or maybe acquisition date instead) and "sale date" make more sense to me for both Real Estate and Securities than "open/close" dates

zakpatterson replied:
You're right. There are some cases where selling happens before buying, but I'll stick to the wording on the 8949 itself, "date acquired" and "date sold or disposed of".

noahm commented on src/components/income/OtherInvestments.tsx:
```diff
@@ -0,0 +1,198 @@
+              <>
+                <LabeledInput
+                  label="Quantity"
+                  patternConfig={Patterns.currency}
```
I don't think Quantity is a currency type? Also, since we're concerning ourselves with quantity we should clarify that the cost of acquisition and sell price are either cost per unit or total amount of the given quantity.

zakpatterson replied:
Yeah I'm thinking ahead to where lots can be split. So if I buy 100 units of x and then later sell 50 units, I need to subdivide the initial purchase block...might be confusing. Fixed format in cc8ecec

noahm commented on src/components/income/OtherInvestments.tsx:
```diff
@@ -0,0 +1,198 @@
+          valueMapping={(x) => x}
+        />
+        <LabeledInput
+          label="Address"
```
"Address" only makes sense for Real Estate. This should just be "Description" or something similar for a Security.

zakpatterson replied:
Ah yes, meant to add a switch there.

noahm reviewed (commented):
Just a few initial thoughts from a quick look. I'd love to have a way to quickly input many transactions but I'll need to think a bit longer on how something like that would need to work.

zakpatterson commented:
Tracking general conception of this feature on the issue: https://github.com/ustaxes/UsTaxes/issues/848

thegrims commented:
Minor nitpick, but I think the alignment of the manage assets text is off
![image](https://user-images.githubusercontent.com/8848557/149890202-45a8c140-8436-4481-93ba-eb169cbe0326.png)

thegrims commented:
![image](https://user-images.githubusercontent.com/8848557/149891795-c70d2982-8b2d-42f7-ac4c-730fea681c31.png)
address form should allow numbers for house numbers

thegrims commented:
Saving real estate gives NaN
![image](https://user-images.githubusercontent.com/8848557/149892118-9dab503b-e36d-40fd-a47c-0981e361c3b2.png)

thegrims commented:
Saving real estate, opening again, asset type is set to stock

thegrims commented:
![image](https://user-images.githubusercontent.com/8848557/149892813-49a227f8-0879-4662-9244-283ad748ace9.png)
Date picker should either not allow sale dates of the non tax year, or show warning that proceeds will not be included in the current tax year filing

zakpatterson commented:
Fixed display issues for Real estate vs other securities, added some styles

![2022-01-19-110412_1477x666_scrot](https://user-images.githubusercontent.com/1987109/150179108-0be6c9d5-0fe4-44f5-8438-1003f514c5fa.png)

zakpatterson commented:
Added warning about closing transactions outside of current year.

![2022-01-19-182518_1244x652_scrot](https://user-images.githubusercontent.com/1987109/150239991-e31c7577-815e-44f0-9f32-ac9129de5a03.png)
